Content.Server/Climbing/ClimbSystem.cs

```diff
@@ -1,121 +1,445 @@
+        }
+        else
+        {
+            var othersMessage = Loc.GetString("comp-climbable-user-climbs-force-other", ("user", args.User),
+                ("moved-user", uid), ("climbable", args.Climbable));
+            args.User.PopupMessageOtherClients(othersMessage);
+
+            var selfMessage = Loc.GetString("comp-climbable-user-climbs-force", ("moved-user", uid),
+                ("climbable", args.Climbable));
+            args.User.PopupMessage(selfMessage);
         }
+    }
 
-        public void ForciblySetClimbing(EntityUid uid, ClimbingComponent? component = null)
+    /// <summary>
+    /// Replaces the current fixtures with non-climbing collidable versions so that climb end can be detected
+    /// </summary>
+    /// <returns>Returns whether adding the new fixtures was successful</returns>
+    private bool ReplaceFixtures(ClimbingComponent climbingComp, PhysicsComponent physicsComp, FixturesComponent fixturesComp)
+    {
+        // Swap fixtures
+        var toAdd = new List<Fixture>();
+        foreach (var (name, fixture) in fixturesComp.Fixtures)
         {
-            if (!Resolve(uid, ref component, false))
-                return;
-            component.IsClimbing = true;
-            UnsetTransitionBoolAfterBufferTime(uid, component);
+            if (climbingComp.Fixtures.Contains(name)
+                || climbingComp.Fixtures.Contains($"{ClimbingFixtureName}-{name}")
+                || fixture.Hard == false
+                || (fixture.CollisionMask & ClimbingCollisionGroup) == 0)
+                continue;
+
+            climbingComp.DisabledFixtures.Add(fixture);
+
+            var climbFixture = new Fixture(fixture.Shape, fixture.CollisionLayer,
+                    fixture.CollisionMask & ~ClimbingCollisionGroup,
+                    true)
+                {ID = $"{ClimbingFixtureName}-{name}"};
+            toAdd.Add(climbFixture);
         }
 
-        private void AddClimbVerb(EntityUid uid, ClimbableComponent component, GetVerbsEvent<AlternativeVerb> args)
+        foreach (var fixture in climbingComp.DisabledFixtures)
         {
-            if (!args.CanAccess || !args.CanInteract || !_actionBlockerSystem.CanMove(args.User))
-                return;
+            _fixtureSystem.DestroyFixture(physicsComp, fixture, manager: fixturesComp);
+        }
+
+        foreach (var fixture in toAdd)
+        {
+            if (!_fixtureSystem.TryCreateFixture(physicsComp, fixture, manager: fixturesComp))
+                return false;
+            climbingComp.Fixtures.Add(fixture.ID);
+        }
 
-            // Check that the user climb.
-            if (!EntityManager.TryGetComponent(args.User, out ClimbingComponent? climbingComponent) ||
-                climbingComponent.IsClimbing)
+        if (!_fixtureSystem.TryCreateFixture(physicsComp,
+            new Fixture(new PhysShapeCircle { Radius = 0.35f }, (int) CollisionGroup.None, ClimbingCollisionGroup, false)
```

Comment:
For `Radius` I went with humanoid fixture size, but it should probably be something more robust. I couldn't think of an easy way to get that.